### ISSUE TITLE
Android: Page doesn't scroll when keyboard is displayed

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#2595]
### Share your bug report, feature request, or comment.

iOS Safari does the correct thing. When you focus on the text field the page scrolls vertically.

With Android everything stays static and the keyboard covers the page, including chat and the text input.

![image](https://user-images.githubusercontent.com/414923/212785139-c40ad996-aff0-4e8e-a743-052a5bac4c11.png)

This is what iOS is doing for comparison: 

![image](https://user-images.githubusercontent.com/414923/212785992-e90f909a-89d0-4af6-8475-6af403ab6191.png)

**Comment by LBBO** [comment]
I can't reproduce this on the current `webv2` branch (392da72c8bc0a3abe572358b905441d984aeb454) using Chrome 109.0.5414.86 on Android 11, see screenshot below. Or am I missing the issue?

![telegram-cloud-photo-size-2-5348131206291636904-y](https://user-images.githubusercontent.com/8076094/212940090-58cf7c5b-3eb7-4482-8680-67504a80eb01.jpg)

**Comment by gabek** [comment]
I'll do some more detailed testing and compare Chrome and Android versions. I'm not sure if it matters in this case, but I was testing against a production build (port 8080, and in my case was https://webv2.owncast.tv), and it looks like you're testing against dev (port 3000). Sometimes production builds act differently unfortunately. 

<blockquote><img src="https://webv2.owncast.tv/thumbnail" width="48" align="right"><div><img src="/img/favicon/favicon-16x16.png" height="14"> Owncast Web v2</div><div><strong><a href="https://webv2.owncast.tv">Owncast Web v2</a></strong></div><div>This is the work in progress Webv2 of Owncast's user interface.</div></blockquote>

**Comment by gabek** [comment]
For some context I'm waiting for Browserstack to give me access to their platform so I can test on more devices/versions. They're giving me a hard time with my account. I'm also on the lookout for a replacement that offers a free open source plan so I can do testing, but I'm kind of stuck right now.

**Comment by gabek** [comment]
This is on:

Google Chrome
Android  11.0
Google Pixel 4A


https://user-images.githubusercontent.com/414923/213090969-6f4f6f2e-77be-41bb-a08c-9b0c0d9e9679.mp4

**Comment by LBBO** [comment]
Just tried it again on local prod with the same result as on dev. I can't try it on webv2.owncast.tv since the stream isn't live at the moment. My device is a OnePlus 7T.

**Comment by andrew-secret** [comment]
> 

Let me know if u find something.

**Comment by LBBO** [comment]
Alright, I tried [USB-debugging Chrome on my device](https://developer.chrome.com/docs/devtools/remote-debugging/) and it seems as if opening the keyboard doesn't change the body's height and just moves it up (which means that it's now possible to scroll in the body 😂 ). When I open the site in the normal Chrome Emulator and change the body's height, I get the exact behavior you're getting. My current theory would be that for some reason your phone changes the body's height while mine just moves it upwards. Could you try to verify that theory?

If this is in fact the case, maybe a solution could be to wait until the page has loaded, check the body's height and force it to have that height forever (until the page is reloaded). If the viewport is changed without a page reload, this would allow us to just move the body upwards manually (for example by having something like `position: absolute; bottom: 0;` on `body`). This would definitely cause issues with #2586 though and seems very hacky overall, so we'll probably want to search for another solution.

Dimensions of `body` without expanded keyboard (note the free space up top that is actually used by the address bar)
![image](https://user-images.githubusercontent.com/8076094/213154688-9a0e1bfb-bf1d-494f-947e-3ed26221fc97.png)
Dimensions of `body` with expanded keyboard (note how body overflows the top edge)
![image](https://user-images.githubusercontent.com/8076094/213154741-3e65087b-f1b5-4218-9cb3-9f3923cdc164.png)

**Comment by gabek** [comment]
> My current theory would be that for some reason your phone changes the body's height while mine just moves it upwards. Could you try to verify that theory?

I'll try and figure out how to answer that. I'm using LambdaTest to test against different devices, so I'll have to look into if I can use developer tools with its remote controlling of devices and emulators.

**Comment by gabek** [comment]
Got it! It does look like the height is changing.

No keyboard:
![image](https://user-images.githubusercontent.com/414923/213800934-6325cf11-0593-4955-baa8-07f05f86b4d4.png)

With keyboard:

![image](https://user-images.githubusercontent.com/414923/213800789-a402a2a6-d333-4401-a260-fdec408c0014.png)

**Comment by LBBO** [comment]
Cool cool cool cool cool. A browser that doesn't behave the same across Android devices. How fun! I'll see if I can figure something out to fix this. Perhaps giving the chat a minimum height and having it stick to the bottom of the screen could fix it.

**Comment by gabek** [comment]
Thank you for looking at it!

**Comment by gabek** [comment]
If it would be helpful feel free to ping me directly on Rocket.Chat or the Fediverse or whatever and I can hook you up with a Lambdatest invite to test against different browsers+devices.

**Comment by LBBO** [comment]
Thanks for the offer to ping you! I hope it won't be necessary since I get at least similar (if not identical) behavior in the chrome device simulator. But if not, I'll definitely reach out!

I've made some progress but am still fighting with the chat input field not being cut off. In my current changes (not pushed anywhere yet as they're not in the least final) remove the `.topSection` element in `Content`. Its styling is applied to all the children via a class that's passed on to them (a bit messy, I know) but this gets rid of the wrapper around the video player. This allows flexbox to properly do its thing and I can now give the chat box a `min-height` and have the video `flex-shrink` in order to grant that `min-height`.

Please let me know if you see any issues with that approach so far. Of course it might be a bit iffy to shrink the video, but I figure that people watching on mobile devices will either be watching the video in full screen mode (i.e. completely without the chat container) or they'll be doing something else (e.g. chatting) and be more interested in that.

If you're ok with this approach, I'll continue my work in that direction. At the moment, the minimum height is somehow being overridden weirdly by exactly the height of the chat input field and I have yet to figure out where that comes from. Also, we need to decide on a minimum height the chat box should have (provided you're ok with my approach).

**Comment by gabek** [comment]
To be completely honest I'm not sure I have any opinion about those specific changes as I'm having a hard time internalizing what is changing just by hearing the class names. So I'd say go for it if you have an approach you'd like to try.

Since you're talking about the video container it might be worthwhile to give #2360 a glance since I've made a few changes over there to try to make the video player size up more what I'd expect it to be. At this point some screen sizes seem ok and some don't. So maybe future work with that should be merged into what you're doing since it does seem to be connected. At the very least make sure you have those changes merged into your branch to make sure it didn't throw anything you're doing off haha.

**Comment by gabek** [comment]
Was throwing around ideas with @gingervitis yesterday and she had a crazy idea that if we knew that there's an on screen keyboard (that may not be possible) and we know the text field has focus we give the top section a top margin of something like `-50%` to scroll it out of the way and make room for the keyboard. It probably has a lot of room for failure, but I wanted to document the idea here.

**Comment by LBBO** [comment]
Sorry for not being as responsive lately, I had alot on my hands with work. Detecting the virtual keyboard sounds pretty tedious tbh and very dependent on the device and browser. There seems to be a new experimental `VirtualKeyboard` API (https://caniuse.com/mdn-api_virtualkeyboard) that could make it easier, but support is pretty bad and I wouldn't really want to deal with it if not necessary.

Thanks for pointing out your other work! I'll make sure to rebase my changes onto it and will hopefully figure out the last remaining quirks Some Time Soon™.

<blockquote><div><strong><a href="https://caniuse.com/mdn-api_virtualkeyboard">VirtualKeyboard API | Can I use... Support tables for HTML5, CSS3, etc</a></strong></div><div>"Can I use" provides up-to-date browser support tables for support of front-end web technologies on desktop and mobile web browsers.</div></blockquote>

**Comment by gabek** [comment]
Oh wow, I'm surprised there's any kind of VirtualKeyboard API, good find. I agree that it's probably not the right direction, though.

**Comment by LBBO** [comment]
Alright, I FINALLY figured out what the issue was (the `.Content_root` was somehow magically getting a `min-height` without it appearing in the dev tools and setting `min-height: 0;` fixed things). Here's a video of the current behavior:

https://user-images.githubusercontent.com/8076094/216779604-9da97f34-940c-4bdd-963d-f53f3b488725.mov

I still need to figure out how to make the play button resize dynamically and the current styling is relatively hard-coded to this exact layout without other banners and stuff, but I should be able to figure that out. It just seems like a good point to check in and make sure this is acceptable behavior in general, since it's not quite the target you described in the issue.

Also, I had to change https://github.com/owncast/owncast/blob/28c9fb11035cfa2e6d57203706b53c6f66081ccb/web/components/video/OwncastPlayer/OwncastPlayer.module.scss#L11 lo `height: 100%` since the video is resizing here. Is that ok? I noticed you introduced the line just a few days ago, so maybe you still remember the specific reason and can tell me off the top of your head if my change would break anything.

**Comment by gabek** [comment]
Very cool! Does it work the same with they on-screen keyboard as it does when you manually resize the window? I would assume it would, but you never know with these things.

As for the play button, I'm not sure if it's worth trying to resize it, because the player is already out of view. The play button seems to scroll vertically with everything else, but not completely, so that's odd. But maybe an option is to just make the play button hide itself in that case, since it should be scrolled out of view anyway. What do you think?

**Comment by LBBO** [comment]
Yeah, it works just as it did before on my phone. You should probably do some more extensive testing once the PR is ready, though, since I don't really have access to many other devices.

My concern regarding the play button is mainly for users in landscape mode, since they'll have a very small screen height and should still be able to interact with the chat and open the video in full screen mode just as before. Ideally, I'll get it to resize with the player, but otherwise I'll probably just hide the overflow.

**Comment by LBBO** [comment]
Alright, styling the play button was a pain in the back because of how video.js is styled by default. I decided to go half-way by making its font size responsive using `clamp`, because anything beyond that would have been digging too deep into video.js's internal structure that might change.

I'm trying to clean up right now and make my changes more robust. In doing so, I noticed that this is what `develop` currently looks like:

![image](https://user-images.githubusercontent.com/8076094/216985861-ad88f3b3-c6f1-4d2b-a949-b0a196bde77a.png)

I'll try to fix that as well, but I just wanted to let you know so you aren't surprised when there are changes that aren't necessarily related to this issue

**Comment by gabek** [comment]
That's so weird! I just tested with an iPhone and it seems to be laying out ok, but my guess is it depends on the size of the window. Thank you for catching that and looking at it on top of everything else!

**Comment by LBBO** [comment]
Alright, I'm finally finished (famous last words). I expect my layout to be usable for screens with a height >= 362px and a width >= 260px (sorry for the arbitrary numbers). Those limits include all the phones I had in the Chrome device emulation dropdown. However, if they're not acceptable for some reason, feel free to let me know. I did as intensive manual testing as I could and didn't find any issues, but I'm not familiar with the full feature set of Owncast, so it might be good if you could make sure everything is still usable at various screen sizes.

Also, do you have a good idea how I could automatically test my changes? Since it's mostly just styling, I'm really unsure how that would work and thus haven't added any tests yet. But I'd be happy to if you could point me in the right direction.

Finally, I noticed that some common phones (e.g. the iPhone 12 Pro) will display the website in desktop mode when used in landscape mode. Is this intended? Otherwise the break points might have to be adjusted accordingly.

**Comment by gabek** [comment]
Awesome, I'll jump in and start checking it out!

> Also, do you have a good idea how I could automatically test my changes?

It kind of depends. We do have Chromatic setup to do UI snapshot testing of individual components, but we keep running out of snapshots per month and it's $150/mo to upgrade, so that's not great. It works to catch changes in components, but not helpful for layout.

We also have cypress tests, but that requires writing actual tests. I've written some, but they're generally sanity tests to make sure things exist that are expected to exist. But it should be trivial to add more, since all the scaffolding is all there, split up between "offline" (not live) and "online" (live) tests. Some are commented out because I never got around to completing them, and some test files are almost empty. You can find what we currently have here: https://github.com/owncast/owncast/tree/develop/test/automated/browser/cypress/e2e. You should be able to run them via [`run.sh`](https://github.com/owncast/owncast/blob/develop/test/automated/browser/run.sh).

> Finally, I noticed that some common phones (e.g. the iPhone 12 Pro) will display the website in desktop mode when used in landscape mode.

Admittedly I haven't put a single thought into landscape until now. I didn't even know people viewed pages in landscape! So it's not on purpose, but if I had to guess what it should do, landscape using a desktop layout sounds like it's more correct than landscape using a mobile (portrait) layout? But I haven't tried it and I'm not sure how bad it is to know for sure. You can see from my two versions of tests (desktop, mobile) that there's no mobile portrait and mobile landscape, so maybe that's a whole other set of tests that need to be written. But I'm also not sure how common it is for anybody to use the site that way. What do you think?

<blockquote><img src="https://repository-images.githubusercontent.com/266445736/4f82fb80-a5b1-11ea-8f1c-69d316f861b5" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/owncast/owncast">owncast/test/automated/browser/cypress/e2e at develop · owncast/owncast</a></strong></div><div>Take control over your live stream video by running it yourself.  Streaming + chat out of the box. - owncast/test/automated/browser/cypress/e2e at develop · owncast/owncast</div></blockquote>

**Comment by LBBO** [comment]
Yeah, I was afraid your answer would be cypress tests. I've only used them for testing behavior and not layouting so far, so I hesitated to use them for that. But I'll take a look at how easy / possible it is tomorrow.

> What do you think?

I would assume that landscape mode is visited at least from time to time when switching to fullscreen mode and back, but all I can really do here is to hope that people use websites just as I do. However, I do think that mobile mode would make more sense here, since it's more likely to be optimized for touch devices and small screens. After all, an old 24" screen might have a similar amount of pixels as a modern iPhone, but the buttons are still way easier to click there with your mouse than on your phone using your fingers. Right now, the only thing I might worry about is the links in the footer (which is definitely complaining on a high level), but stuff may change in the future.

Finally, just as an FYI: I found something that I missed in the desktop layout, so you'll want to pull my recent changes.

**Comment by gabek** [comment]
Yeah, I'm not actually sure how or if Cypress could work for testing visual layouts. That's what Storybook+Chromatic is good for. There probably is a route forward for automatically testing layout changes, but it would take some investment.

- Make the layouts operate as standalone components, just like any other component. That way they can be tested outside of the full web app. I don't know how possible this is without rewriting a lot and adding complexity around the layout components.
- We create storybook stories for the layouts so they can be viewed in Storybook.
- We reduce the number of snapshots being made right now so we don't run out of test runs every month on Chromatic.
- And then Chromatic would report on major changes to layouts just like it would with any other component.

Having the layouts treated like any other standalone component probably is the more _correct_ thing to do, but there's just a lot of variables there since they're kind of treated as the top level at this point. So we'd have to build another level and wrap the layouts in data providers of some sort. I think it'd be a lot of work.

On top of that, I wish there was a competitor to Chromatic that we could use as an alternative service, but there doesn't seem to be any other players in the space, and I guess that's why Chromatic can charge $150/mo for their smallest plan.

**Comment by gabek** [comment]
> However, I do think that mobile mode would make more sense here, since it's more likely to be optimized for touch devices and small screens. 

It does look like on a mobile device using landscape it not rendering great using the mobile layout just because the mobile layout is optimized for a single column tall screen. It has very little vertical space to fit anything, but maybe I'm misinterpreting what you're saying with the Landscape + Mobile layout.

![image](https://user-images.githubusercontent.com/414923/217122183-a3348808-0695-4f12-aec3-5e936a6f31f2.png)
![image](https://user-images.githubusercontent.com/414923/217122386-eddadc86-5803-4de6-8a26-bbacc118873a.png)

To compare, this is what the old webv1 looks like on landscape using the desktop layout. It does seem to fit things nicer side by side.
![image](https://user-images.githubusercontent.com/414923/217122890-c716266c-b52d-4cea-b257-b327856503c8.png)

**Comment by LBBO** [comment]
I see your point about vertical and horizontal space. My concern is just that desktop layouts don't always care that much about large clickable surfaces and touch-friendliness. More importantly, I could see new contributors seeing a Desktop layout and then add features which aren't that usable without a mouse because they assume it's safe. Essentially, it's just important to keep in mind that the desktop breakpoint currently might be used on mobile phones.

**Comment by LBBO** [comment]
Since I'm not familiar with Chromatic: would it be possible to achieve something similar by running some cypress screenshot tests (https://docs.cypress.io/guides/tooling/visual-testing) on a local storybook instance? Something like "open storybook, click here and there to navigate to a story, take a screenshot and compare it with the last version". It wouldn't be pretty, but it wouldn't be as limited (at least not in the same way).

If you'd like, I can try and write some screenshot tests for this layout. It might be difficult to make them reproducible for live streams, but I could try. But I can also definitely understand if you think it's too much hassle and to messy for too little payoff.

<blockquote><div><strong><a href="https://docs.cypress.io/guides/tooling/visual-testing">Visual Testing | Cypress Documentation</a></strong></div><div>What you'll learn</div></blockquote>

**Comment by gabek** [comment]
Good news! I got Chromatic to move us to the Open Source plan that includes 35,000 snapshots a month for free. So they should start running on PRs again now. I went ahead and sent you a link to it on the Fediverse.

**Comment by gabek** [comment]
> My concern is just that desktop layouts don't always care that much about large clickable surfaces and touch-friendliness.

That makes sense. I'm also not sure how good the mobile layout we have is, either. In theory it should be small screen and touch friendly, but falls mostly into the "good enough" territory just because we haven't had anybody to focus on mobile design (or any design, actually!).

A part of this redesign and rebuild was to improve mobile support, but we've been just trying to make it up as we go along since we haven't had a designer to help on that front. So that's why now that landscape has been brought up I don't know if there's any good answer. The screenshot I took shows that mobile layout + landscape is unusable because there's not enough vertical room for the video + chat. The real answer is probably to have a third layout that's specifically for mobile + landscape, but as you've probably seen we're having enough problems just with two layouts (desktop, mobile + portrait).

**Comment by LBBO** [comment]
Yeah, a completely new layout sounds like way too much work and really annoying for the future. I would probably recommend some sort of mobile-first approach for future development that then causes everything to be touch-friendly by default and things would only change (for desktop) where necessary. With that in place, I'd say landscape should be good just using the normal desktop layout (at least for now).

Of course, enforcing something like that is difficult, especially in OS. Perhaps a definition of done could be added somehow (either to issues or to pull requests) that could at least notify people of the policy and have them check off boxes for different layouts that they tested, but some might consider that a bit intrusive and it might be a bad idea for other reasons.

Thanks for the invitation to Chromatic! I'll definitely take a look around. Does this mean you'd like me to add tests for the layouts? If so, is there anything specific that you want / don't want?

**Comment by gabek** [comment]
I do think all of this stems from not having a single design direction up front. Common sense says we should know what things should look like in all layouts before they're built, but right now everything is kind of made up as we go along. There's no standard set of rules because everyone is thinking completely different, or not thinking about it at all. The fact that landscape hasn't been even thought about until now means there's a gap in the process. But I've been unable to fill that gap unfortunately. I've been trying for years now to find design help, and it hasn't worked out, but I'm not giving up! We really need somebody to help with these kind of layout discussions.

Luckily with Storyboard + Chromatic no tests need to be written. It's just like you mentioned previously, comparing snapshots of components. As long as a component has a Storybook story it'll automatically be tested.

The downside of Chromatic is you need to be a member of the organization to see the test results, so that's why I made sure to add you since we were talking about snapshot tests.
